feat(aio): force live-example to embedded-style on narrow screen

wardbell opened the pull request:
## Updated

@IgorMinar asked that "mobile" screen constraint be determined by viewport size, not user-agent.

After further experimentation, I determined that, while the regular style plunker is useless on narrow screens, the _embedded-style_  plunker is functional which means that we _can show plunkers on mobile devices_.

Therefore, this PR's 2nd commit changes the logic. It forces plunkers to _embedded style_ on narrow (1000px) screens. It allows download of sample zips (I suppose you might want to email it from your phone).

It removes the user-agent test for mobile devices from the first commit. It keeps the refactoring of window resize in the `DeviceService` because that check is needed in two places and because it seems like a favorable abstraction for future use.

@sjtrimble CSS changes to control the embedded image width on narrow screens is more valid than ever.

### Original

Cannot run plunkers on mobile devices nor does it make sense to download code sample to such devices.  Therefore, should detect when on mobile and tell user that the live example is not available.

PR adds `DeviceService` to detect mobile device. The check does not consider a tablet (e.g., iPad) a mobile device because I believe that plunkers should display fine on them. While it is foolish to download the code to a tablet, we do not prevent it.

`DeviceService` also listen for window resize events. Moved window resize detection from `AppComponent` to `DeviceService.displayWidth` _observable_  so others can hear about it from a shared, device-focused service.

I thought this would be important for addressing the no-LiveExample-on-mobile requirement. It turns out it wasn't needed but I kept it there because it feels like `AppComponent` should delegate this responsibility to a service that knows about the device.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

wardbell committed:
feat(aio): detect mobile device and tell user live-example is not available

Adds DeviceService to detect mobile device and listen for window resize events.

mary-poppins commented:
The angular.io preview for 9adc212 is available [here][1].

[1]: https://pr15887-9adc212080c902725379f47ef96ec302d798dff0.ngbuilds.io/

gkalpak commented on aio/src/app/shared/device.service.ts:
```diff
@@ -0,0 +1,41 @@
+  constructor() {
+    this.isMobile = this.isMobileCheck();
+
+    if (window) {
```
I am not thrilled to access window directly here, but not a blocking issue 🌊

wardbell replied:
Do you have a suggestion?

gkalpak replied:
In another PR, I used a [`Global` provider][1] so that I can more easily/cleanly [mock it][2] [in tests][3]. But maybe I am overthinking it. I am too used to AngularJS' `$window` :grin: 

[1]: https://github.com/gkalpak/angular/blob/316374091aea53ff90b302f5e5645279fd479f59/aio/src/app/sw-updates/global.value.ts
[2]: https://github.com/gkalpak/angular/blob/316374091aea53ff90b302f5e5645279fd479f59/aio/src/app/sw-updates/sw-update-notifications.service.spec.ts#L30
[3]: https://github.com/gkalpak/angular/blob/316374091aea53ff90b302f5e5645279fd479f59/aio/src/app/sw-updates/sw-update-notifications.service.spec.ts#L175-L179

gkalpak commented on aio/src/app/embedded/live-example/live-example.component.spec.ts:
```diff
@@ -254,12 +257,36 @@ describe('LiveExampleComponent', () => {
+        expect(liveExampleDe.nativeElement.innerText).toContain('not available on mobile');
+      });
+    }));
+  });
```
We should have a test for both embedded and non-embedded plnkrs.

wardbell replied:
Added

gkalpak commented on aio/src/app/embedded/live-example/live-example.component.spec.ts:
```diff
@@ -254,12 +257,36 @@ describe('LiveExampleComponent', () => {
+  });
 });
+
+//// helpers ////
```
Can we please move all the helpers in one place?

wardbell replied:
Done

gkalpak reviewed (changes requested):
No tests for `DeviceService` ? 😱

wardbell commented on aio/src/app/shared/device.service.ts:
```diff
@@ -0,0 +1,41 @@
+  readonly sideBySideWidth = 1032;
+
+  isMobile = false;
+  displayWidth = new ReplaySubject<number>();
```
I frankly don't know how to test `DeviceService` itself because it interacts directly with the browser environment and I have no idea how to simulate that. I'm open to ideas.

Note that we never tested this interaction before my refactor. Thanks to the `DeviceService`, we can simulate the effects of different devices on the application code w/o actually changing devices.

My bad. Should be `new ReplaySubject<number>(1);` .   Fixing.

mary-poppins commented:
The angular.io preview for 8ff736f is available [here][1].

[1]: https://pr15887-8ff736f8b437b31aa7c560f0ed2a495b4dabb23b.ngbuilds.io/

gkalpak commented on aio/src/app/shared/device.service.ts:
```diff
@@ -0,0 +1,41 @@
+
+    // http://stackoverflow.com/questions/11381673/detecting-a-mobile-browser
+    // tslint:disable-next-line:curly max-line-length whitespace
+    (function(a){if(/(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/i.test(a)||/1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac| ...
```
Wow! OK, I will trust StackOverflow on this one :grin:

IgorMinar replied:
why do we use useragent string for this? it's the viewport resolution that matters not if it's a mobile device. Can we just use that instead?

wardbell replied:
@IgorMinar I thought about that. It is the device and not the viewport that determines whether the link should show.

On desktop I might have the browser window narrow. I can always expand it to allow the plunker to play. I can also download the code. Neither of these options are available on mobile.

IgorMinar replied:
Please use viewport size. Only developers will see the site in narrow mode on desktop.

wardbell replied:
On it

gkalpak commented on aio/src/app/shared/device.service.ts:
```diff
@@ -0,0 +1,41 @@
+    (function(a){if(/(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/i.test(a)||/1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac| ...
+
+    return check;
+  };
```
Unnecessary `;`

wardbell replied:
fixed

gkalpak commented on aio/src/app/shared/device.service.ts:
```diff
@@ -0,0 +1,41 @@
+    }
+  }
+
+  isMobileCheck () {
```
Unnecessary space between `isMobileCheck` and `()` :grin:

wardbell replied:
Fixed

gkalpak reviewed (approved):
A couple of nitpicks.
Regarding the `DeviceService` tests: Not a blocker (especially since we didn't test this before), but there is some logic in `DeviceService` that could be tested.
(An injectable `global/window` would indeed come in handy here.)

mary-poppins commented:
The angular.io preview for 9d304ce is available [here][1].

[1]: https://pr15887-9d304ce7ea83d0b612370b0aaef456415093795b.ngbuilds.io/

wardbell committed:
feat(aio): force plunker to embedded-style on narrow (mobile) screens

Regular plunker is unusable on narrow screen
Refactors LiveExampleComponent and adds tests.
Refactor width detection to `DeviceService` because need to know width change in 2 places.
Keep “disable” option add in earlier spikes because simple and potentially useful in future.

mary-poppins commented:
The angular.io preview for fba26fb is available [here][1].

[1]: https://pr15887-fba26fb71c089bb5a85b1ebc849b61d0f9e923ee.ngbuilds.io/

wardbell commented:
Reworked the logic to depend on viewport width and eliminate user-agent check. See revised initial PR comment which explains everything.

gkalpak commented on aio/src/app/embedded/live-example/live-example.component.spec.ts:
```diff
@@ -144,6 +158,22 @@ describe('LiveExampleComponent', () => {
+      });
+    }));
+
+    it('should be flat style even when flatstyle requested', async(() => {
```
"even when"?

wardbell replied:
Oops. Copy/paste error. Just "when ... requested"

wardbell replied:
The "rules" are described in the documentation at the top.

gkalpak commented on aio/src/app/embedded/live-example/live-example.component.ts:
```diff
@@ -63,56 +68,73 @@ const zipBase = 'content/zips/';
+  readonly narrowWidth = 1000;
 
+  attrs: any;
+  destroy = new Subject();
```
We are calling this `onDestroy` in `ApiService` and `SwUpdatesService`. Let's be consistent :grin:

wardbell replied:
K

gkalpak commented on aio/src/app/embedded/live-example/live-example.component.ts:
```diff
@@ -63,56 +68,73 @@ const zipBase = 'content/zips/';
+        const isFlatStyle = boolFromAtty(flatStyle);
+
+        const embeddedStyle = this.getAttrValue(['embedded-style', 'embeddedstyle', 'embeddedStyle']);
+        const isEmbeddedStyle = width < this.narrowWidth || boolFromAtty(embeddedStyle, !isFlatStyle);
```
In this block, we know that `width > this.narrowWidth`.

wardbell replied:
Good catch

gkalpak reviewed (changes requested):
Can you write up the rules of when each plnkr style is applied (and whether it should be opened in a new tab etc). I am pretty confused atm 😕

wardbell commented on aio/src/app/embedded/live-example/live-example.component.spec.ts:
```diff
@@ -254,12 +284,38 @@ describe('LiveExampleComponent', () => {
+    }));
+
+    it('should be embedded style even when flatstyle requested', async(() => {
+      setHostTemplate('<live-example flatstyle></live-example>');
```
It can be any of the aliased choices - "flatstyle", "flat-style", or "flatStyle" - known to appear in the docs. I didn't feel it was worth testing them all.

wardbell commented on aio/src/app/embedded/live-example/live-example.component.ts:
```diff
@@ -63,56 +68,73 @@ const zipBase = 'content/zips/';
 
-    let exampleDir = attrs.name;
+    const attrs = this.attrs = this.getAttrs();
+    let exampleDir = this.attrs.name;
```
k

mary-poppins commented:
The angular.io preview for aced5e0 is available [here][1].

[1]: https://pr15887-aced5e09c0516c570f3dd22878da659b58fab82c.ngbuilds.io/

mary-poppins commented:
The angular.io preview for d00565b is available [here][1].

[1]: https://pr15887-d00565b27589f867a3f705ea5c787f4a871a4cf6.ngbuilds.io/
